### ISSUE TITLE
listenHeadChanges ChainNotify call failed: RPC client error: sendRequest failed: websocket routine exiting

### ISSUE THREAD

**Issue by cryptowhizzard** [filecoin-project/lotus#4816]
**Describe the bug**
2020-11-12T07:18:08.665 INFO filecoin_proofs::api > generate_piece_commitment:finish
2020-11-12T07:18:08.674 INFO filecoin_proofs::api > generate_piece_commitment:start
2020-11-12T07:18:08.711Z	INFO	events	events/events.go:104	restarting listenHeadChanges
2020-11-12T07:18:08.711Z	ERROR	events	events/events.go:95	listen head changes errored: listenHeadChanges ChainNotify call failed: RPC client error: sendRequest failed: websocket routine exiting
2020-11-12T07:18:08.711Z	INFO	events	events/events.go:104	restarting listenHeadChanges
2020-11-12T07:18:08.711Z	ERROR	events	events/events.go:95	listen head changes errored: listenHeadChanges ChainNotify call failed: RPC client error: sendRequest failed: websocket routine exiting

**To Reproduce**
Running latest master code

**Version (run `lotus version`):**
Latest master of november 10 2020.

**Additional context**

- I have reported this problem twice, but it is still not fixed. The miner crashes randomly. 

- I am running miner as user lotus.

- Output of ulimit -n :

lotus@main:~$ ulimit -n
1000000

- The machine is running daemon + miner + PC1.

**Comment by jennijuju** [comment]
its hard to tell which commit is `latest master`. And have you checked your node status? how was it syncing? you may need to provide more information before we can investigate.

**Comment by cryptowhizzard** [comment]
The setup is as follows :

1 Machine ( Chain daemon + PC1 + Unseal )
5 workers connected for PC2 / C2 

The first machine crashes , however lotus daemon and lotus-miner keeps running. It does not pledge new sectors however / do POST or other jobs. The workers stay connected to it , but they don't do any work.

The only way to resolve is to restart the lotus-miner proces.

I know that pinpointing is hard , but by indicating that i am on latest master you can rule out "resolved things" prior version 1.1.2
Version 1.1.2 also had this problem here.

**Comment by shownakamoto** [comment]
cc @magik6k 
I noticed websocket is not stable (and Its connection recovering mechanism) and this issue happened a lot in our cluster.
Just curious do you have any refactoring plan, like making jsonrpc based on some middleware like ZeroMQ?

**Comment by askender** [comment]
I have the same issue today, in lotus 1.1.2
Daemon:  1.1.2+git.d4cdc6d3+api0.17.0
Local: lotus version 1.1.2+git.d4cdc6d3

other logs:
```

2020-11-16T00:18:37.189+0800    ERROR   rpc     go-jsonrpc@v0.1.2-0.20201008195726-68c6a2704e49/websocket.go:654        Connect
ion timeout      {"remote": "127.0.0.1:1234"}

2020-11-16T00:44:33.238+0800	ERROR	rpc	go-jsonrpc@v0.1.2-0.20201008195726-68c6a2704e49/websocket.go:497	sending ping message: write tcp 127.0.0.1:60122->127.0.0.1:1234: use of closed network connection

2020-11-16T00:44:37.226+0800	ERROR	storageminer	storage/wdpost_sched.go:96	ChainNotify error: RPC client error: sendRequest failed: websocket routine exiting
2020-11-16T00:44:38.238+0800	ERROR	rpc	go-jsonrpc@v0.1.2-0.20201008195726-68c6a2704e49/websocket.go:497	sending ping message: write tcp 127.0.0.1:60122->127.0.0.1:1234: use of closed network connection
```

**Comment by jennijuju** [comment]
duplicate $4762

**Comment by jennijuju** [comment]
@cryptowhizzard nope, what i meant is that we appreciate your report, however, our master head changes frequently and in fact, we have released v1.1.3. And thats is why i asked for your `lotus version` to get the exact commit you were on.
